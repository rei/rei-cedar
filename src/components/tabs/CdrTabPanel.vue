<script setup lang="ts">
import {
  computed, ref, inject, useCssModule, watch,
} from 'vue';
import kebabCase from '../../utils/kebabCase';
import { selectedTabKey } from './symbols';

defineOptions({
  name: 'CdrTabPanel',
});

const props = defineProps({
  /** Sets reference identifier for tab content. This property is required and is necessary for accessibility. Must be unique for each tabPanel, and cannot be the same as the `aria-labelledby` property. */
  id: String,
  /** Sets tab display name. Required and must be unique for each tab. If `id` is not provided, this value will be used as the reference identifier. */
  name: String,
  /** Sets reference identifier for tab header. This property is required and is necessary for accessibility. Must be unique for each tabPanel, and cannot be the same as the `id` property. */
  ariaLabelledby: String,
});

const emits = defineEmits({
  /**
   * Emits when active tab is changed
   * @params state, panelId
   */
    'tab-change': null,
});

const style = useCssModule();
const selectedTabName = inject(selectedTabKey, ref(null));
const isActive = computed(() => props.name === selectedTabName?.value);
const panelId = computed(() => props.name ? `${kebabCase(props.name)}-panel` : undefined);

watch(isActive, (state) => {
  emits('tab-change', state, panelId.value);
});
</script>

<template>
  <section
    v-show="isActive"
    tabindex="0"
    :aria-hidden="!isActive"
    role="tabpanel"
    :class="style['cdr-tab-panel']"
    :id="panelId"
    :aria-labelledby="name"
  >
    <!-- @slot CdrTabPanel content -->
    <slot />
  </section>
</template>

<style lang="scss" module src="./styles/CdrTabPanel.module.scss">
</style>
