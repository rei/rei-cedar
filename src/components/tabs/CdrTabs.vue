<script setup lang="ts">
import {
  ref, provide, onMounted, nextTick, computed, useCssModule, useSlots,
} from 'vue';
import type { ComponentInternalInstance } from 'vue';
import { debounce } from '../../utils/debounce'
import {
  CdrColorBackgroundPrimary, CdrSpaceOneX, CdrSpaceHalfX,
} from '@rei/cdr-tokens/dist/rei-dot-com/js/cdr-tokens.mjs';
import mapClasses from '../../utils/mapClasses';
import { modifyClassName } from '../../utils/buildClass';
import { selectedTabKey } from './symbols';

/** Organizes related content into groups for people to navigate between */
defineOptions({
  name: 'CdrTabs'
});

const props = defineProps({
  /**
   * Sets height of the tabs container element.
   * Passing a `px` value will render tabs with a static height,
   * passing `auto` will render tabs with variable height based on content size.
   */
    height: {
    type: String,
    default: '240px',
  },
  /** Sets the index of the tab that should be active on initial page load. Note that this property is zero-indexed. */
  activeTab: {
    type: Number,
    default: 0,
  },
  /**
   * Modifies the style variants for this component
   * @values centered, compact, full-width, no-border
   */
  modifier: String,
  /**
   * Use `small` to reduce spacing around the tabs for a denser visual design
   * @demoSelectMultiple true
   * @values small
  */
  size: String,
  /**
   * Sets the background color of the tab.
   * For CdrTabs that are rendered on non-primary backgrounds.
   * Pass the background-color into the component to ensure that the scrolling gradients render correctly.
   */
  backgroundColor: {
    type: String,
    default: CdrColorBackgroundPrimary,
  },
});
const style = useCssModule();
const slots = useSlots();

const slottedTabs = computed<any>(() => slots.default?.()[0]?.children?.length
  ? slots.default?.()[0]?.children
  : slots.default?.());

const baseClass = 'cdr-tabs';

// Refs
// const tabs = ref(slottedTabs?.map((tab) => ({
//   name: tab.props.name,
//   disabled: tab.props.disabled,
//   id: tab.props['aria-labelledby'],
// })));

const tabs = computed(() => {
  if (slottedTabs.value) {
    return slottedTabs.value.map((tab: ComponentInternalInstance) => ({
      name: tab.props.name,
      disabled: tab.props.disabled,
      id: tab.props['aria-labelledby'],
    }));
  } else {
    return []; // Return an empty array if there are no slottedTabs
  }
});

const selectedTabName = ref(null);
const selectedIndex = ref<number | null>(null);
// const selectedIndex = ref(0);
const headerOverflow = ref(false);
const headerWidth = ref(0);
const tabElements = ref<HTMLButtonElement[]>([]);
const tablist = ref<HTMLUListElement | null>(null);
const containerEl = ref<HTMLDivElement | null>(null);
const overflowLeft = ref(false);
const overflowRight = ref(false);
const underlineOffsetX = ref(0);
const underlineWidth = ref(0);

provide(selectedTabKey, selectedTabName);

// Computed
const modifierClass = computed(() => props.modifier
  ? modifyClassName('cdr-tabs', props.modifier) : '');
const sizeClass = computed(() => props.size ? modifyClassName('cdr-tabs', props.size) : '');
const underlineStyle = computed(() => ({
  transform: `translateX(${underlineOffsetX.value}px)`,
  width: `${underlineWidth.value}px`,
}));
const gradientLeftStyle = computed(() => {
  const gradient = `linear-gradient(to left, rgba(255, 255, 255, 0), ${props.backgroundColor})`;
  return {
    background: gradient,
  };
});
const gradientRightStyle = computed(() => {
  const gradient = `linear-gradient(to right, rgba(255, 255, 255, 0),
    ${props.backgroundColor})`;
  return {
    background: gradient,
  };
});
const checkIfActive = (index: number, tab: any) => (selectedIndex.value === index && !tab.disabled);
const calculateOverflow = () => {
  let containerWidth = 0;
  if (containerEl.value) {
    containerWidth = containerEl.value.offsetWidth;
  }
  headerOverflow.value = headerWidth.value > containerWidth;
  if (headerOverflow.value) {
    // Get Scroll Position
    const scrollX = tablist.value?.scrollLeft ?? 0;
    overflowLeft.value = scrollX > 1;
    overflowRight.value = (scrollX + 1) < (headerWidth.value - containerWidth);
    
  } else {
    overflowLeft.value = false;
    overflowRight.value = false;
  }
};

const getHeaderWidth = () => {
  let headerElements: Element[] = [];
  if (tablist.value) {
    headerElements = Array.from(tablist.value.children);
  }
  let totalWidth = 0;
  headerElements.forEach((element, i) => {
    // account for margin-left on header elements
    if (i > 0) {
      totalWidth += props.size === 'small' ? Number(CdrSpaceHalfX) : Number(CdrSpaceOneX);
    }
    totalWidth += (element as HTMLElement).offsetWidth || 0;
  });
  return totalWidth;
};

const updateUnderline = () => {
  if (tabElements.value.length > 0) {
    const activeTab = tabElements.value[selectedIndex.value || 0];
    const activeRect = (activeTab as HTMLElement).getBoundingClientRect();
    const parentRect = tablist.value?.getBoundingClientRect();
    const offset = parentRect ? activeRect.x - parentRect.x : 0;

    underlineOffsetX.value = tablist.value ? offset - tablist.value?.scrollLeft : 0;
    underlineWidth.value = activeRect.width;

    // shrink/hide the underline if it scrolls outside the container
    if (parentRect && underlineOffsetX.value + underlineWidth.value >= parentRect.width) {
      underlineWidth.value = Math.max(0, parentRect.width - underlineOffsetX.value);
      underlineOffsetX.value = Math.min(underlineOffsetX.value, parentRect.width);
    } else if (underlineOffsetX.value < 0) {
      underlineWidth.value = Math.max(0, underlineWidth.value + underlineOffsetX.value);
      underlineOffsetX.value = 0;
    }
  }
};
const selectTab = async (index: number) => {
  const tabToSelect = tabElements.value[index];
  selectedTabName.value = tabs.value[index].name;
  selectedIndex.value = index;
  await nextTick();
  (tabToSelect as HTMLElement).focus();
  updateUnderline();
};

const selectTabNext = () => {

  const isLastTab = (selectedIndex.value === tabElements.value.length - 1);
  if (isLastTab) {
    return;
  }

  let nextIndex = (selectedIndex.value || 0) + 1;
  if (tabElements.value[nextIndex].disabled) {
    nextIndex += 1;
  }

  const nextIndexExists = (nextIndex <= tabElements.value.length - 1);
  if (!nextIndexExists) {
    return;
  }

  selectTab(nextIndex);
};

const selectTabPrev = () => {
  const isFirstTab = (selectedIndex.value && selectedIndex.value <= 0);
  if (isFirstTab) {
    return;
  }

  let prevIndex = (selectedIndex.value || 0) - 1;
  if (tabElements.value[prevIndex].disabled) {
    prevIndex -= 1;
  }

  const previousIndexExists = (prevIndex >= 0);
  if (!previousIndexExists) {
    return;
  }
  selectTab(prevIndex);
};

const setInitialTabStates = () => {
  tabElements.value.forEach((tab, index) => {
    if (!tab.disabled && selectedIndex.value === null) {
      selectedIndex.value = index;
      selectedTabName.value = tabs.value[index].name;
    }
  });
};

onMounted(() => {
  setInitialTabStates();
  headerWidth.value = getHeaderWidth();
  calculateOverflow();
  setTimeout(() => {
    updateUnderline();
  }, 250);
  window.addEventListener('resize', debounce(() => {
    headerWidth.value = getHeaderWidth();
    calculateOverflow();
    updateUnderline();
  }, 250));
  tablist.value?.addEventListener('scroll', debounce(() => {
    calculateOverflow();
    updateUnderline();
  }, 50));
});


</script>

<template>
  <div
    :class="mapClasses(style, baseClass, modifierClass, sizeClass)"
    ref="containerEl"
    :style="{ height }"
  >
    <div
      :class="style['cdr-tabs__gradient-container']"
    >
      <div
        :class="mapClasses(
          style,
          'cdr-tabs__gradient',
          'cdr-tabs__gradient--left',
          overflowLeft ? 'cdr-tabs__gradient--active' : ''
        )"
        :style="gradientLeftStyle"
      />
      <ul
        :class="style['cdr-tabs__header-container']"
        role="tablist"
        ref="tablist"
      >
        <li
          v-for="(tab, index) in tabs"
          :key="`${tab.name}-${index}`"
          role="presentation"
          :class="style['cdr-tabs__header']"
        >
          <button
            :ref="(el: HTMLButtonElement | any) => { tabElements[index] = el }"
            :id="tab.id"
            :disabled="tab.disabled"
            :aria-selected="checkIfActive(index, tab)"
            :tabIndex="checkIfActive(index, tab) ? 0 : -1"
            :class="mapClasses(
              style,
              checkIfActive(index, tab) ? 'cdr-tabs__header-item-active' : '',
              'cdr-tabs__header-item',
              tab.disabled ? 'cdr-tabs__header-item--disabled' : '',
            )"
            role="tab"
            @click.prevent="selectTab(index)"
            @keyup.right="selectTabNext"
            @keyup.left="selectTabPrev"
          >
            {{ tab.name }}
          </button>
        </li>
      </ul>
      <div
        :class="mapClasses(
          style,
          'cdr-tabs__gradient',
          'cdr-tabs__gradient--right',
          overflowRight ? 'cdr-tabs__gradient--active' : '',
        )"
        :style="gradientRightStyle"
      />
      <div
        :class="style['cdr-tabs__underline']"
        :style="underlineStyle"
      />
    </div>
    <!-- @slot CdrTabs content (CdrTabPanel components) -->
    <slot />
  </div>
</template>

<style lang="scss" module src="./styles/CdrTabs.module.scss">
</style>