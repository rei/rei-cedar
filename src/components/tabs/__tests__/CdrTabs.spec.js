import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrTabs from '../CdrTabs.vue';
import CdrTabPanel from '../CdrTabPanel.vue';
import { h } from 'vue';
// Tests use nextTick because of the nextTick in mounted hook of tabs

describe('CdrTabs', () => {
  let wrapper;
  beforeEach(()=>{
    wrapper = mount(CdrTabs, {
      slots: {
        default: [
          h(CdrTabPanel, {name: 'tab1', 'aria-labelledby': 'tab-small-one',}), 
          h(CdrTabPanel, {name: 'tab2'}),
          h(CdrTabPanel, {name: 'tab3', disabled: true}),
          h(CdrTabPanel, {name: 'tab4'})
        ],
      },
    });
  })
  describe('mounted', () => {
    it('mounts tabs', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('mounts with cdr-tab-panel children', async () => {
      const spyCalculateOverflow = vi.spyOn(wrapper.vm, 'calculateOverflow');
      const spyUpdateUnderline = vi.spyOn(wrapper.vm, 'updateUnderline');

    //   await wrapper.vm.$nextTick();

      expect(wrapper.vm.tabs.length).toBe(4);
      expect(wrapper.findAll('button').length).toBe(4);
      expect(wrapper.vm.selectedIndex).toBe(0);
      setTimeout(() => { // for debounce
        expect(spyCalculateOverflow).toHaveBeenCalled();
        expect(spyUpdateUnderline).toHaveBeenCalled();
      }, 500);
    });
  });

  describe('event listeners', () => {
    it('handles scroll event', async () => {
      const spyCalculateOverflow = vi.spyOn(wrapper.vm, 'calculateOverflow');
      const spyUpdateUnderline = vi.spyOn(wrapper.vm, 'updateUnderline');

      wrapper.find('.cdr-tabs__header-container').element.dispatchEvent(new Event('scroll'));
      await wrapper.vm.$nextTick();

      setTimeout(() => { // for debounce
        expect(wrapper.vm.overflowLeft).toBe(false);
        expect(spyCalculateOverflow).toHaveBeenCalled();
        expect(spyUpdateUnderline).toHaveBeenCalled();
      }, 600);
    });

    it('handles resize event', async () => {
      const spyCalculateOverflow = vi.spyOn(wrapper.vm, 'calculateOverflow');
      const spyUpdateUnderline = vi.spyOn(wrapper.vm, 'updateUnderline');
      const spyGetHeaderWidth = vi.spyOn(wrapper.vm, 'getHeaderWidth');
      window.dispatchEvent(new Event('resize'));
      await wrapper.vm.$nextTick();

      setTimeout(() => { // for debounce
        expect(spyGetHeaderWidth).toHaveBeenCalledTimes(2);
        expect(spyCalculateOverflow).toHaveBeenCalledTimes(2);
        expect(spyUpdateUnderline).toHaveBeenCalledTimes(2);
      }, 500);
    });
  });

  it('getNextTab and getPreviousTab', async() => {
    expect(wrapper.vm.selectedIndex).toBe(0);
    wrapper.vm.selectTabNext();
    expect(wrapper.vm.selectedIndex).toBe(1);
    wrapper.vm.selectTabNext();
    //TODO: There should be a separate it block to test that disabled tabs are skipped
    expect(wrapper.vm.selectedIndex).toBe(3);
    wrapper.vm.selectTabPrev();
    expect(wrapper.vm.selectedIndex).toBe(1);
    wrapper.vm.selectTabPrev();
    expect(wrapper.vm.selectedIndex).toBe(0);
  });

  describe('handleClick', () => {
    it('changes the selectedIndex', async () => {
      await wrapper.vm.$nextTick();
      expect(wrapper.vm.selectedIndex).toBe(0);
      wrapper.findAll('button')[1].trigger('click');
      await wrapper.vm.$nextTick();
      expect(wrapper.vm.selectedIndex).toBe(1);
    });
  });

  it('handles right arrow key', async () => {
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedIndex).toBe(0);
    // Trigger right arrow keyup event
    wrapper.findAll('button')[0].trigger('keyup.right');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedIndex).toBe(1);
  });

  it('handles left arrow key', async () => {
    wrapper.vm.selectedIndex = 1;
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedIndex).toBe(1);
    // Trigger left arrow keyup event
    wrapper.findAll('button')[1].trigger('keyup.left');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedIndex).toBe(0);
  });

  describe('overflow classes', () => {
    it('adds gradient-left class', async () => {
      await wrapper.vm.$nextTick();
      wrapper.vm.overflowLeft = true;
      wrapper.vm.overflowRight = false;
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.cdr-tabs__gradient--right.cdr-tabs__gradient--active').exists()).toBe(false);
      expect(wrapper.find('.cdr-tabs__gradient--left.cdr-tabs__gradient--active').exists()).toBe(true);
    });

    it('adds gradient-right class', async () => {
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.cdr-tabs__gradient--right.cdr-tabs__gradient--active').exists()).toBe(true);
      expect(wrapper.find('.cdr-tabs__gradient--left.cdr-tabs__gradient--active').exists()).toBe(false);
    });
  });

  it('accessibility', async () => {
    const tab1 = wrapper.find('#tab-small-one');
    //TODO: Separate describe block for accessibility with single it blocks for each of these expect statements
    expect(tab1.attributes()['aria-selected']).toBe('true');
    expect(tab1.attributes()['role']).toBe('tab');
    expect(wrapper.find('.cdr-tabs__header').attributes()['role']).toBe('presentation');
    expect(wrapper.find('.cdr-tabs__header-container').attributes()['role']).toBe('tablist');
  });

  it('focuses on the selected tab button element', async () => {
    await wrapper.vm.selectTab(1);
    setTimeout(()=>{
      expect(wrapper.findAll('button')[1]).toBe(document.activeElement);
    }, 500);
  });

  it('calculateOverflow', async () => {
    // wrapper.setData({ headerWidth: 2000 });
    await wrapper.vm.$nextTick();
    wrapper.vm.calculateOverflow();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.headerOverflow).toBe(true);
  });
});