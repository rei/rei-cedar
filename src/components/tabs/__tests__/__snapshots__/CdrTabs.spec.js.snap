// Vitest Snapshot v1

exports[`CdrTabs > mounted > mounts tabs 1`] = `
<div
  class="cdr-tabs"
  style="height: 240px;"
>
  <div
    class="cdr-tabs__gradient-container"
  >
    <div
      class="cdr-tabs__gradient cdr-tabs__gradient--left"
    />
    <ul
      class="cdr-tabs__header-container"
      role="tablist"
    >
      
      <li
        class="cdr-tabs__header"
        role="presentation"
      >
        <button
          aria-selected="true"
          class="cdr-tabs__header-item-active cdr-tabs__header-item"
          id="tab-small-one"
          role="tab"
          tabindex="0"
        >
          tab1
        </button>
      </li>
      <li
        class="cdr-tabs__header"
        role="presentation"
      >
        <button
          aria-selected="false"
          class="cdr-tabs__header-item"
          role="tab"
          tabindex="-1"
        >
          tab2
        </button>
      </li>
      <li
        class="cdr-tabs__header"
        role="presentation"
      >
        <button
          aria-selected="false"
          class="cdr-tabs__header-item cdr-tabs__header-item--disabled"
          disabled=""
          role="tab"
          tabindex="-1"
        >
          tab3
        </button>
      </li>
      <li
        class="cdr-tabs__header"
        role="presentation"
      >
        <button
          aria-selected="false"
          class="cdr-tabs__header-item"
          role="tab"
          tabindex="-1"
        >
          tab4
        </button>
      </li>
      
    </ul>
    <div
      class="cdr-tabs__gradient cdr-tabs__gradient--right cdr-tabs__gradient--active"
    />
    <div
      class="cdr-tabs__underline"
      style="transform: translateX(0px); width: 0px;"
    />
  </div>
  <!-- @slot CdrTabs content (CdrTabPanel components) -->
  
  <section
    aria-hidden="false"
    aria-labelledby="tab1"
    class="cdr-tab-panel"
    id="tab1-panel"
    role="tabpanel"
    style=""
    tabindex="0"
  >
    <!-- @slot CdrTabPanel content -->
    
    
  </section>
  <section
    aria-hidden="true"
    aria-labelledby="tab2"
    class="cdr-tab-panel"
    id="tab2-panel"
    role="tabpanel"
    style="display: none;"
    tabindex="0"
  >
    <!-- @slot CdrTabPanel content -->
    
    
  </section>
  <section
    aria-hidden="true"
    aria-labelledby="tab3"
    class="cdr-tab-panel"
    disabled="true"
    id="tab3-panel"
    role="tabpanel"
    style="display: none;"
    tabindex="0"
  >
    <!-- @slot CdrTabPanel content -->
    
    
  </section>
  <section
    aria-hidden="true"
    aria-labelledby="tab4"
    class="cdr-tab-panel"
    id="tab4-panel"
    role="tabpanel"
    style="display: none;"
    tabindex="0"
  >
    <!-- @slot CdrTabPanel content -->
    
    
  </section>
  
</div>
`;
